<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-[Chapter-09]-추천-시스템" data-toc-modified-id="1.-[Chapter-09]-추천-시스템-1">1. [Chapter 09] 추천 시스템</a></span><ul class="toc-item"><li><span><a href="#1.1-추천-시스템의-개요와-배경" data-toc-modified-id="1.1-추천-시스템의-개요와-배경-1.1">1.1 추천 시스템의 개요와 배경</a></span></li><li><span><a href="#1.2-콘텐츠-기반-필터링-추천-시스템" data-toc-modified-id="1.2-콘텐츠-기반-필터링-추천-시스템-1.2">1.2 콘텐츠 기반 필터링 추천 시스템</a></span></li><li><span><a href="#1.3-최근접-이웃-협업-필터링" data-toc-modified-id="1.3-최근접-이웃-협업-필터링-1.3">1.3 최근접 이웃 협업 필터링</a></span></li><li><span><a href="#1.4-잠재-요인-협업-필터링" data-toc-modified-id="1.4-잠재-요인-협업-필터링-1.4">1.4 잠재 요인 협업 필터링</a></span><ul class="toc-item"><li><span><a href="#1.4.1-행렬-분해의-이해" data-toc-modified-id="1.4.1-행렬-분해의-이해-1.4.1">1.4.1 행렬 분해의 이해</a></span></li><li><span><a href="#1.4.2-확률적-경사-하강법(SGD)을-이용한-행렬-분해" data-toc-modified-id="1.4.2-확률적-경사-하강법(SGD)을-이용한-행렬-분해-1.4.2">1.4.2 확률적 경사 하강법(SGD)을 이용한 행렬 분해</a></span></li></ul></li><li><span><a href="#1.5-콘텐츠-기반-필터링-실습---TMDB-5000-영화-데이터-세트" data-toc-modified-id="1.5-콘텐츠-기반-필터링-실습---TMDB-5000-영화-데이터-세트-1.5">1.5 콘텐츠 기반 필터링 실습 - TMDB 5000 영화 데이터 세트</a></span><ul class="toc-item"><li><span><a href="#1.5.1-데이터-로딩-및-가공" data-toc-modified-id="1.5.1-데이터-로딩-및-가공-1.5.1">1.5.1 데이터 로딩 및 가공</a></span></li><li><span><a href="#1.5.2-장르-콘텐츠-유사도-측정" data-toc-modified-id="1.5.2-장르-콘텐츠-유사도-측정-1.5.2">1.5.2 장르 콘텐츠 유사도 측정</a></span></li><li><span><a href="#1.5.3-장르-콘텐츠-필터링을-이용한-영화-추천" data-toc-modified-id="1.5.3-장르-콘텐츠-필터링을-이용한-영화-추천-1.5.3">1.5.3 장르 콘텐츠 필터링을 이용한 영화 추천</a></span></li></ul></li><li><span><a href="#1.6-아이템-기반-최근접-이웃-협업-필터링-실습" data-toc-modified-id="1.6-아이템-기반-최근접-이웃-협업-필터링-실습-1.6">1.6 아이템 기반 최근접 이웃 협업 필터링 실습</a></span><ul class="toc-item"><li><span><a href="#1.6.1-데이터-가공-및-변환" data-toc-modified-id="1.6.1-데이터-가공-및-변환-1.6.1">1.6.1 데이터 가공 및 변환</a></span></li></ul></li><li><span><a href="#1.7-행렬-분해를-이용한-잠재-요인-협업-필터링-실습" data-toc-modified-id="1.7-행렬-분해를-이용한-잠재-요인-협업-필터링-실습-1.7">1.7 행렬 분해를 이용한 잠재 요인 협업 필터링 실습</a></span></li><li><span><a href="#1.8-파이썬-추천-시스템-패키지---Surprise" data-toc-modified-id="1.8-파이썬-추천-시스템-패키지---Surprise-1.8">1.8 파이썬 추천 시스템 패키지 - Surprise</a></span></li></ul></li></ul></div>

# 1. [Chapter 09] 추천 시스템

## 1.1 추천 시스템의 개요와 배경

- 추천 시스템의 유형 : 콘텐츠 기반 필터링 방식, 협업 필터링 방식(최근접 이웃 협업 필터링, 잠재 요인 협업 필터링)

초창기에는 콘텐츠 기반이나 최근접 이웃 협업 필터링이 주로 사용됐지만, 현재 대부분의 온라인 스토어에서 잠재 요인 협업 필터링 적용

개인화 특성을 좀 더 강화하기 위해 하이브리드 형식으로 콘텐츠 기반과 협업 기반을 적절히 결합하는 경우도 있음

## 1.2 콘텐츠 기반 필터링 추천 시스템

- **콘텐츠 기반 필터링 방식 : 사용자가 선호하는 아이템과 비슷한 콘텐츠를 가진 다른 아이템을 추천하는 방식** ex) 비슷한 장르, 감독, 배우

## 1.3 최근접 이웃 협업 필터링

- 협업 필터링 방식 : 사용자가 평가한 아이템을 기반으로 아직 평가하지 않은 아이템의 예측 평가를 도출하여 추천하는 방식

최근접 이웃 방식과 잠재 요인 방식 모두 사용자-아이템 평점 행렬 데이터에만 의지해 추천을 수행

사용자-아이템 평점 행렬의 행은 개별 사용자, 열은 개별 아이템, 값은 평점으로 구성되며, **pivot_table()** 함수로 형태 변경

일반적으로 많은 아이템을 열로 가지는 다차원 행렬이며, 사용자가 평점을 매기는 경우가 많지 않아 희소 행렬 특성을 가짐

최근접 이웃 협업 필터링은 메모리 협업 필터링이라고도 하며, 사용자 기반과 아이템 기반으로 나눌 수 있음

- **사용자 기반 최근접 이웃 방식 : 특정 사용자와의 유사도가 높은 TOP-N 사용자가 좋아하는 아이템을 추천하는 방식**

(Customers like you also bought these items.)


- **아이템 기반 최근접 이웃 방식 : 사용자들의 평가 척도가 유사한 아이템을 추천하는 방식**

(Customers who bought this item also bought these items.)

일반적으로 사용자 기반보다는 아이템 기반 협업 필터링의 정확도가 높기 때문에 **대부분 아이템 기반의 알고리즘.** 주로 코사인 유사도 이용

## 1.4 잠재 요인 협업 필터링

- **잠재 요인 협업 필터링 방식 : 사용자-아이템 평점 행렬 속의 잠재 요인을 추출해 추천하는 방식**

대규모 다차원 행렬을 차원 감소 기법으로 분해하는 과정에서 잠재 요인을 추출하는 **행렬 분해** 기반의 잠재 요인 협업 필터링

사용자-아이템 행렬 데이터를 사용자-잠재 요인 행렬과 잠재 요인-아이템 행렬로 분해한 후,

분해된 두 행렬의 내적을 통해 새로운 예측 사용자-아이템 평점 행렬 데이터를 만들어 아이템에 대한 예측 평점을 생성

(잠재 요인이 정확히 어떤 것인지 알 수 없으나, 영화 평점 기반의 사용자-아이템 행렬에서 장르별 특성 선호도로 해석할 수 있음)

### 1.4.1 행렬 분해의 이해

- 행렬 분해 : 다차원의 매트릭스를 저차원 매트릭스로 분해하는 기법 ex) SVD, NMF

SVD는 NaN 값이 없는 행렬에만 적용할 수 있으므로, 주로 확률적 경사 하강법 방식을 이용해 SVD 수행

### 1.4.2 확률적 경사 하강법(SGD)을 이용한 행렬 분해

P와 Q 행렬로 계산된 **예측 R 행렬 값이 실제 R 행렬 값과 최소의 오류를 가질 수 있도록 비용 함수 최적화를 통해 P와 Q를 유추함**

1) P와 Q를 임의의 행렬로 설정

2) P와 Q.T 값을 곱해 예측 R 행렬을 계산하고, 실제 R 행렬과의 오류 값 계산

3) 오류 값을 최소화하도록 P와 Q 행렬을 적절한 값으로 업데이트

4) 2,3번 작업을 반복하면서 P와 Q 값을 업데이트하여 근사화

## 1.5 콘텐츠 기반 필터링 실습 - TMDB 5000 영화 데이터 세트

anaconda 창에서 kaggle datasets download -d tmdb/tmdb-movie-metadata 입력하여 데이터 다운로드

### 1.5.1 데이터 로딩 및 가공

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

movies = pd.read_csv('./data/tmdb-movie-metadata/tmdb_5000_movies.csv')
print(movies.shape)
movies.head(3)

(4803, 20)


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466

In [2]:
df = movies[['id','title','genres','vote_average','vote_count','popularity','keywords','overview']]

In [3]:
pd.set_option('max_colwidth', 100)
df[['genres','keywords']][:1]

genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...   

                                                                                              keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...

genres 칼럼에서 딕셔너리의 key인 'name'으로 개별 장르를 추출할 수 있으며, keywords 칼럼 역시 동일한 구조를 가짐

ast 모듈의 **literal_eval()** 함수를 이용하면 문자열을 list [dict1, dict2] 객체로 변환할 수 있음

In [4]:
from ast import literal_eval

df['genres'] = df['genres'].apply(literal_eval)
df['keywords'] = df['keywords'].apply(literal_eval)

In [5]:
df['genres'] = df['genres'].apply(lambda x: [y['name'] for y in x])
df['keywords'] = df['keywords'].apply(lambda x: [y['name'] for y in x])
df[['genres', 'keywords']][:1]

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                                                                              keywords  
0  [culture clash, future, space war, space colony, society, space travel, futuristic, romance, spa...

### 1.5.2 장르 콘텐츠 유사도 측정

1) 문자열로 변환된 genres 칼럼을 Count 기반으로 피처 벡터화 변환

2) 변환한 데이터 세트를 코사인 유사도를 통해 비교. 이를 위해 데이터 세트 레코드별로 타 레코드와 장르에서 유사도 값을 가지는 객체 생성

3) 장르 유사도가 높은 영화 중 평점이 높은 순으로 영화 추천

In [6]:
from sklearn.feature_extraction.text import CountVectorizer

# CountVectorizer를 적용하기 위해 공백문자로 word 단위가 구분되는 문자열로 변환
df['genres_literal'] = df['genres'].apply(lambda x: (' ').join(x))
count_vec = CountVectorizer(min_df=0, ngram_range=(1,2))
genre_mat = count_vec.fit_transform(df['genres_literal'])
print(genre_mat.shape)

(4803, 276)


In [8]:
from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[:1])

(4803, 4803)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]]


**argsort()[:, ::-1]**을 이용해 높은 순으로 정렬된 비교 행 위치 인덱스 값을 가져오고 0번 값만 샘플로 추출

In [10]:
genre_sim_sorted_idx = genre_sim.argsort()[:, ::-1]
print(genre_sim_sorted_idx[:1])

[[   0 3494  813 ... 3038 3037 2401]]


0번 레코드의 경우 3494번 레코드, 813번 레코드 순으로 유사도가 높음

### 1.5.3 장르 콘텐츠 필터링을 이용한 영화 추천

In [11]:
# 장르 유사도에 따라 영화를 추천하는 함수 생성
def find_sim_movie(df, sorted_idx, title_name, top_n=10):
    title_movie = df[df['title'] == title_name]
    
    title_idx = title_movie.index.values
    similar_idx = sorted_idx[title_idx, :(top_n)]
    
    print(similar_idx)
    similar_idx = similar_idx.reshape(-1)     # 1차원 array로 변경
    
    return df.iloc[similar_idx]

In [12]:
similar_movies = find_sim_movie(df, genre_sim_sorted_idx, 'The Godfather', 10)
similar_movies[['title', 'vote_average']]

[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


title  vote_average
2731                          The Godfather: Part II           8.3
1243                                    Mean Streets           7.2
3636                                   Light Sleeper           5.7
1946  The Bad Lieutenant: Port of Call - New Orleans           6.0
2640         Things to Do in Denver When You're Dead           6.7
4065                                      Mi America           0.0
1847                                      GoodFellas           8.2
4217                                            Kids           6.8
883                              Catch Me If You Can           7.7
3866                                     City of God           8.1

영화 평점 정보인 vote_average 칼럼을 이용해 평점에 따라 필터링해서 최종 추천하는 방식으로 변경

그 전에 voto_average는 왜곡된 데이터를 가지고 있으므로, 이를 먼저 확인

In [13]:
df[['title','vote_average','vote_count']].sort_values('vote_average', ascending=False)[:10]

title  vote_average  vote_count
3519          Stiff Upper Lips          10.0           1
4247     Me You and Five Bucks          10.0           2
4045     Dancer, Texas Pop. 81          10.0           1
4662            Little Big Top          10.0           1
3992                 Sardaarji           9.5           2
2386            One Man's Hero           9.3           2
2970        There Goes My Baby           8.5           2
1881  The Shawshank Redemption           8.5        8205
2796     The Prisoner of Zenda           8.4          11
3337             The Godfather           8.4        5893

생소한 영화들이 평가 횟수는 매우 작고 평점은 높은 왜곡된 데이터를 가지고 있음

IMDB에서 사용하는 평가 횟수에 대한 가중치가 부여된 평점 방식을 이용해 새롭게 평점 부여 (C: 전체 영화의 평균 평점, m: 최소 투표 횟수)

In [19]:
C = df['vote_average'].mean()
m = df['vote_count'].quantile(0.6)     # 투표 횟수를 상위 60% 기준으로 정함
print('C: ', round(C, 3), 'm: ', round(m, 3))

C:  6.092 m:  370.2


In [20]:
def weighted_vote_average(record):
    v = record['vote_count']
    R = record['vote_average']
    
    return ( (v/(v+m)) * R ) + ( (m/(m+v)) * C ) 

df['weighted_vote'] = df.apply(weighted_vote_average, axis=1)

In [21]:
df[['title','vote_average','weighted_vote','vote_count']].sort_values('weighted_vote', ascending=False)[:10]

title  vote_average  weighted_vote  vote_count
1881  The Shawshank Redemption           8.5       8.396052        8205
3337             The Godfather           8.4       8.263591        5893
662                 Fight Club           8.3       8.216455        9413
3232              Pulp Fiction           8.3       8.207102        8428
65             The Dark Knight           8.2       8.136930       12002
1818          Schindler's List           8.3       8.126069        4329
3865                  Whiplash           8.3       8.123248        4254
809               Forrest Gump           8.2       8.105954        7927
2294             Spirited Away           8.3       8.105867        3840
2731    The Godfather: Part II           8.3       8.079586        3338

In [22]:
# 새롭게 정의된 평점 기준에 따라 영화를 추천하는 함수 생성
def find_sim_movie(df, sorted_idx, title_name, top_n=10):
    title_movie = df[df['title'] == title_name]
    title_idx = title_movie.index.values
    
    # top_n의 2배에 해당하는 장르 유사성이 높은 인덱스 추출
    similar_idx = sorted_idx[title_idx, :(top_n * 2)]
    similar_idx = similar_idx.reshape(-1)
    # 기준 영화 인덱스는 제외
    similar_idx = similar_idx[similar_idx != title_idx]
    
    return df.iloc[similar_idx].sort_values('weighted_vote', ascending=False)[:top_n]

similar_movies = find_sim_movie(df, genre_sim_sorted_idx, 'The Godfather', 10)
similar_movies[['title', 'vote_average', 'weighted_vote']]

title  vote_average  weighted_vote
2731       The Godfather: Part II           8.3       8.079586
1847                   GoodFellas           8.2       7.976937
3866                  City of God           8.1       7.759693
1663  Once Upon a Time in America           8.2       7.657811
883           Catch Me If You Can           7.7       7.557097
281             American Gangster           7.4       7.141396
4041              This Is England           7.4       6.739664
1149              American Hustle           6.8       6.717525
1243                 Mean Streets           7.2       6.626569
2839                     Rounders           6.9       6.530427

## 1.6 아이템 기반 최근접 이웃 협업 필터링 실습

https://grouplens.org/datasets/movielens/latest/ (ml-latest-small)

### 1.6.1 데이터 가공 및 변환

In [24]:
import pandas as pd
import numpy as np

movies = pd.read_csv('./data/ml-latest-small/movies.csv')
ratings = pd.read_csv('./data/ml-latest-small/ratings.csv')
print(movies.shape)
print(ratings.shape)
movies.head(3)

(9742, 3)
(100836, 4)


movieId                    title  \
0        1         Toy Story (1995)   
1        2           Jumanji (1995)   
2        3  Grumpier Old Men (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance

In [25]:
ratings.head(3)

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224

**pivot_table()** 함수를 이용해 사용자를 로우로, 영화를 칼럼으로 구성한 데이터 세트로 변경

In [26]:
ratings = ratings[['userId', 'movieId', 'rating']]
ratings_matrix = ratings.pivot_table('rating', index='userId', columns='movieId')
ratings_matrix.head(3)

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     NaN     4.0     NaN     NaN     4.0     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  193583  193585  193587  193609  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  

[3 rows x 9724 columns]

In [27]:
# 칼럼명을 영화명으로 변경하기 위해 movies와 조인
rating_movies = pd.merge(ratings, movies, on='movieId')

# columns='title'로 title 칼럼으로 피벗 수행
ratings_matrix = rating_movies.pivot_table('rating', index='userId', columns='title')

# 평점이 NaN인 값을 모두 0으로 변환
ratings_matrix = ratings_matrix.fillna(0)
ratings_matrix.head(3)

title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              0.0                                      0.0   
2              0.0                                      0.0   
3              0.0                                      0.0   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                          0.0                  0.0   
2                          0.0                  0.0   
3                          0.0                  0.0   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                             0.0                              0.0   
2                             0.0                              0.0   
3                             0.0                              0.0   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                      0.0                   0.0                          0.0   
2                      0.0                   0.0                          0.0   
3                      0.0                   0.0                          0.0   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                                  0.0  ...          0.0           0.0   
2                                  0.0  ...          0.0           0.0   
3                                  0.0  ...          0.0           0.0   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1                 0.0                      0.0   
2                 0.0                      0.0   
3                 0.0                      0.0   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                          0.0   
2                                                          0.0   
3                                                          0.0   

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1                   0.0         0.0                             0.0   
2                   0.0         0.0                             0.0   
3                   0.0         0.0                             0.0   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                         4.0                                        0.0  
2                         0.0                                        0.0  
3                         0.0                                        0.0  

[3 rows x 9719 columns]

## 1.7 행렬 분해를 이용한 잠재 요인 협업 필터링 실습

## 1.8 파이썬 추천 시스템 패키지 - Surprise